[SYCL] Move splitting module code into a new file

mlychkov opened the pull request:
sycl-post-link tool's code should be grouped into several source files
to separate support of different functionalities, provided by the tool.
It should improve code readability and sustainability.
Current part of sycl-post-link tool code reorganization:
- Move splitting mechanism into a separate file.
- Make several ModuleSplitter classes to reduce comparisons during
  splitting.
- Modify types that handle module splitting to simplify migration to
  multithreading split in the future.

Signed-off-by: Mikhail Lychkov <mikhail.lychkov@intel.com>

psamolysov-intel commented on llvm/tools/sycl-post-link/ModuleSplitter.cpp:
```diff
@@ -0,0 +1,433 @@
+      *M, SplitEsimd, EmitOnlyKernelsAsEntryPoints);
+  bool DoSplit = (Groups.size() > 1);
+
+  std::unique_ptr<ModuleSplitterBase> Splitter;
+  if (!DoSplit)
+    Splitter.reset(new ModuleCopier(std::move(M), std::move(Groups)));
+  else
+    Splitter.reset(new ModuleSplitter(std::move(M), std::move(Groups)));
+  return Splitter;
```
```suggestion
  if (DoSplit)
    return std::make_unique<ModuleSplitter>(std::move(M), std::move(Groups));
  else
    return std::make_unique<ModuleCopier>(std::move(M), std::move(Groups));
```

psamolysov-intel commented on llvm/tools/sycl-post-link/ModuleSplitter.cpp:
```diff
@@ -0,0 +1,433 @@
+  bool DoSplit = (Mode != SPLIT_NONE &&
+                  (Groups.size() > 1 || !Groups.cbegin()->Functions.empty()));
+
+  std::unique_ptr<ModuleSplitterBase> Splitter;
+  if (!DoSplit)
+    Splitter.reset(new ModuleCopier(std::move(M), std::move(Groups)));
+  else
+    Splitter.reset(new ModuleSplitter(std::move(M), std::move(Groups)));
+  return Splitter;
```
```suggestion
  if (DoSplit)
    return std::make_unique<ModuleSplitter>(std::move(M), std::move(Groups));
  else
    return std::make_unique<ModuleCopier>(std::move(M), std::move(Groups));
```

psamolysov-intel commented on llvm/tools/sycl-post-link/ModuleSplitter.cpp:
```diff
@@ -0,0 +1,433 @@
+  if (DeviceGlobals)
+    checkImageScopedDeviceGlobals(*M, Groups);
+  bool DoSplit = (Mode != SPLIT_NONE &&
+                  (Groups.size() > 1 || !Groups.cbegin()->Functions.empty()));
```
We are checking for at least a single function in the *first* group but why not in *every* one?

psamolysov-intel commented on llvm/tools/sycl-post-link/ModuleSplitter.cpp:
```diff
@@ -0,0 +1,433 @@
+  decltype(GVs.size()) I = 0;
+  while (I != GVs.size()) {
+    const auto *F = cast<Function>(GVs[I++]);
+    for (const auto &I : instructions(F))
```
This `I` shadows the first `I` what is an index. Maybe we should rename the first `I` in `Var` or `Idx`, or something like this?

psamolysov-intel commented on llvm/tools/sycl-post-link/ModuleSplitter.cpp:
```diff
@@ -0,0 +1,433 @@
+        // be controlled by an option.
+        error("no '" + Twine(ATTR_SYCL_MODULE_ID) +
+              "' attribute for entry point '" + F.getName() +
+              "', per-module split not possible");
```
```suggestion
              "', per-module split is not possible");
```

psamolysov-intel commented on llvm/tools/sycl-post-link/sycl-post-link.cpp:
```diff
@@ -616,9 +324,10 @@ void saveModuleIR(Module &M, StringRef OutFilename) {
 }
 
-void saveModuleProperties(Module &M, const EntryPointGroup &ModuleEntryPoints,
+void saveModuleProperties(module_split::ModuleDesc &ResM,
```
Do we change `ModuleDesc` here? If not (I can find where), should the reference be `const`?

mlychkov replied:
We don't change it, but `MAM.getResult<SYCLKernelParamOptInfoAnalysis>(Module &M)` is called here which uses standard `ModuleAnalysisManager` interface.

psamolysov-intel commented on llvm/tools/sycl-post-link/sycl-post-link.cpp:
```diff
@@ -616,9 +324,10 @@ void saveModuleIR(Module &M, StringRef OutFilename) {
+void saveModuleProperties(module_split::ModuleDesc &ResM,
                           const GlobalBinImageProps &ImgPSInfo,
                           const std::string &PropSetFile) {
+  Module &M = ResM.getModule();
```
```suggestion
  const Module &M = ResM.getModule();
```

psamolysov-intel reviewed (commented):
Good job, thank you @mlychkov Some my comments are below. I'm not sure we should use nested namespaces in the `sycl-post-link` tool, it is not such big as the compiler driver, for example, but this is just an opinion.

mlychkov commented on llvm/tools/sycl-post-link/Utils.h:
```diff
@@ -0,0 +1,28 @@
+//===----------- Utils.h - utilities for sycl-post-link tool --------------===//
```
Thank you. I just had older version of sycl-post-link code and didn't see that new file.
Moved `Utils.h` content into `Support.h` file.

mlychkov commented on llvm/tools/sycl-post-link/ModuleSplitter.h:
```diff
@@ -0,0 +1,104 @@
+
+  Module &getInputModule() { return *InputModule; }
+  std::unique_ptr<Module> releaseInputModule() {
+    return std::move(InputModule);
```
Added assert check for methods accessing `InputModule` field.

mlychkov commented on llvm/tools/sycl-post-link/sycl-post-link.cpp:
```diff
@@ -108,7 +96,7 @@ cl::opt<bool> SplitEsimd{"split-esimd",
 // TODO Design note: sycl-post-link should probably separate different kinds of
 // its functionality on logical and source level:
-//  - LLVM IR module splitting
+//  - LLVM IR module splitting (done)
```
Removed the line.

psamolysov-intel reviewed (dismissed):
LGTM

mlychkov commented:
> I'm not sure we should use nested namespaces in the `sycl-post-link` tool, it is not such big as the compiler driver, for example, but this is just an opinion.

I have no strong opinion here regarding using nested namespaces or not. @AlexeySachkov what do you think?

mlychkov commented:
> @mlychkov Why have you removed the check for the correct using of device global variables?

Sorry, I removed it by accident. Added that check back.

mlychkov commented:
> I'm not sure the namespace prefix is required before a class, the class is already a "namespace" for its methods. So I see no such prefixes in `AliasAnalysis.cpp`, for example: `AAResults::AAResults(AAResults &&Arg)`, not `llvm::AAResults::AAResults(AAResults &&Arg)`. As I understand, the directive `using namespace llvm` takes all the reachable classes from the `llvm` namespace into the scope and lets us write the class name without the namespace prefix even in the `ClassName::method()` constructs (as well as we write class names for parameter and variable declarations, etc.)
> 
> As I get, the namespace prefix is required for free function definitions only.

Yes, thank you for prompt argumentation. I am agree with it. Removed llvm namespace prefix from member functions definitions.

kbobrovs commented on llvm/tools/sycl-post-link/ModuleSplitter.h:
```diff
@@ -0,0 +1,110 @@
+  size_t totalSplits() const { return Groups.size(); }
+
+  // Check that there are no more submodules to split.
+  bool EOL() const { return GroupsIt == Groups.cend(); }
```
Nit: name acronym for "end of line/life", neither seem to fit the purpose. `hasMoreSplits`?

mlychkov replied:
Thank you for suggestion. Applied.

kbobrovs reviewed (dismissed):
@mlychkov, just wanted to note that my invoke_simd patches depend on this refactoring, as one of the patches makes changes to splitting and post-link. None of my review comments are supposed to hold this PR, and can be addressed later.

mlychkov commented on llvm/tools/sycl-post-link/ModuleSplitter.cpp:
```diff
@@ -0,0 +1,428 @@
+      return Scope_Global;
+    }
+
+    if (hasIndirectFunctionCalls(M))
```
There is already a comment inside that function.

mlychkov commented on llvm/tools/sycl-post-link/ModuleSplitter.cpp:
```diff
@@ -0,0 +1,428 @@
+    // are also considered as entry points (except __spirv_* and __sycl_*
+    // functions)
+    return F.hasFnAttribute(ATTR_SYCL_MODULE_ID) &&
+           !isSpirvSyclBuiltin(F.getName());
```
I think it should be applied in an additional patch and have LIT tests for that cases.

mlychkov replied:
Made an issue for that improvement: https://github.com/intel/llvm/issues/5878

mlychkov commented on llvm/tools/sycl-post-link/ModuleSplitter.cpp:
```diff
@@ -0,0 +1,428 @@
+
+    while (!Workqueue.empty()) {
+      const User *U = Workqueue.back();
+      Workqueue.pop_back();
```
Thank you

mlychkov commented on llvm/tools/sycl-post-link/ModuleSplitter.cpp:
```diff
@@ -0,0 +1,428 @@
+  }
+}
+
+void collectFunctionsToExtract(SetVector<const GlobalValue *> &GVs,
```
It is just for interface unification. CloneModule uses vector of `const GlobalValue *` values.

mlychkov commented on llvm/tools/sycl-post-link/ModuleSplitter.cpp:
```diff
@@ -0,0 +1,428 @@
+      if (const auto *CB = dyn_cast<CallBase>(&I))
+        if (const Function *CF = CB->getCalledFunction())
+          if (!CF->isDeclaration())
+            GVs.insert(CF);
```
Added a comment that this is not expected.

psamolysov-intel commented on llvm/tools/sycl-post-link/ModuleSplitter.cpp:
```diff
@@ -0,0 +1,427 @@
+
+} // namespace
+
+bool llvm::module_split::ModuleDesc::isEsimd() {
```
Just a nit: it looks like the `llvm` namespace prefix can be removed here too in the same way as you did it for the `module_split::getSlitterByKernelType` function.

mlychkov commented:
Failures in OCL GEN9 and ESIMD Emu test suites are known XPASSes.
Failure in precommit test suite is infrastructure issue. Re-run has passed: http://icl-jenkins.sc.intel.com:8080/blue/organizations/jenkins/SYCL_CI%2Fintel%2FWin%2FTest_Suite/detail/Test_Suite/9916/pipeline/

psamolysov-intel reviewed (approved):
LGTM

mlychkov commented:
Failed tests in ESIMD emulator:
  SYCL :: accessor_gather_scatter.cpp
  SYCL :: accessor_load_store.cpp
  SYCL :: api/esimd_pack_unpack_mask.cpp
  SYCL :: api/simd_copy_to_from.cpp
had failed in another PRs as well. However, they no more failed for the last 2 days.
I restarted tests to check if they are passed here as well.
